src/util/std_types.h

```diff
@@ -40,6 +40,17 @@ class bool_typet:public typet
   }
 };
 
+/// The int type. 
+/// This is necessary for the type checking of loop variants, which should have
+/// some kind of integer type.
+class int_typet:public typet
```

Comment:
See
https://github.com/diffblue/cbmc/blob/0ae6010a4e3977e591d9f4dd6029f6cfc6dc4508/src/util/mathematical_types.h#L20

Reply:
Thank you so much for your suggestion! Ideally, I want to use a "bounded" int type (as opposed to an unbounded int type) because every loop variant/ranking function must be bounded below. Although `ID_signed_int` is used in CBMC's code base, it doesn't seem to be a valid type of GOTO programs. That is, the function `implicit_typecast` in https://github.com/diffblue/cbmc/blob/0ae6010a4e3977e591d9f4dd6029f6cfc6dc4508/src/ansi-c/c_typecheck_typecast.cpp#L13 doesn't accept `ID_signed_int` as the target of type-casting. Please let me know if you have a suggestion!

Reply:
In CPROVER, Integer generally means mathematically bounded integer and C `int` is a `ID_signedbv` or a `ID_unsignedbv`.  The choice between the two has significant impact in terms of decidability and performance.

Do you need to fix the type of this annotation?  I am /guessing/ what you are going to do is take the expression `dec(x,y,z,...)` and generate two assertions : `dec(x,y,z,...) < dec(__CPROVER_old(x), __CPROVER_old(y), ... )` and `0 <= dec(x,y,z,...)`?  If so then you can do this with any ordered type that has a 0.

Reply:
Thanks a lot for your swift response! As you remarked, one idea is to bound loop variants from below; i.e. `0 <= dec(x,y,z,...)`. Another idea is to use an integer type that is already bounded below (e.g. `ID_signedbv` and `ID_unsignedbv`). I was initially considering the first idea, and then switched to the second idea. However, maybe the first idea is better.

Reply:
By the way, I finished implementing a prototype of loop variants. If I use the type `integer_typet` from https://github.com/diffblue/cbmc/blob/0ae6010a4e3977e591d9f4dd6029f6cfc6dc4508/src/util/mathematical_types.h#L20, I can successfully run `goto-cc toy.c -o toy.goto` and `goto-instrument --enforce-all-contracts toy.goto toy.2.goto`. However, `cbmc toy.2.goto` fails - it returns an error message of UNIMPLEMENTED. Instead, if I use `ID_signedbv`, `cbmc toy.2.goto` successfully runs. So I probably should switch from `integer_typet` to a different integer type.

Reply:
From a decision procedures point of view, bit-vectors (`ID_signedbv`, `ID_unsignedbv`, `ID_bv`) and mathematical integers (`ID_integer`) are very very different things.  For a fixed bit-width, bit-vector decision problems are in NP.  There are even some sub-theories that are in P.  Mathematical integers are in undecidable in the general case and NP in the restricted linear case.  The decision procedures you use for them are very different and converting between them in an expression is generally very expensive because it is hard for the decision procedures to exchange information in a useful way (because they work so very differently).  The native back-end does not support more than trivial reasoning over mathematical integers.

My suggestion would be to avoid fixing which you are using unless you really have to.  What you are doing should work just fine for both.  If you just have it as an expression then the user can give a ranking function in `ID_signedbv`, `ID_integer`, `ID_float`, whatever.  You only have to block the case of `ID_real` or add something like `2*(dec(x'',y'',z'') - dec(x',y',z')) < (dec(x',y',z') - dec(x,y,z))` where each `'` denotes an previous iteration.

Reply:
It is feasible to specify/construct ranking functions using bitvector types. 2LS does that, for instance. However, you need to be careful to use sufficiently wide types so as to avoid any overflows in the arithmetic occurring in the expressions involved.

Reply:
Thanks for the pointers, Martin, Peter. For now we implemented `implicit_typecast_arithmetic` check on the ranking function. This check is quite liberal and avoids fixing the expected type for the ranking function.

In the next PR, when we instrument the code to generate the monotonic decrement assertions from ranking functions, we would carefully explore the various corner cases that you have pointed out.